File: apps/desktop/src/features/trade/futuresRoots.ts
/** Known tradable futures roots (FuturesRoots.swift). */
export const KNOWN_FUTURES_ROOTS = ['MES', 'ES', 'MNQ', 'NQ', 'CL', 'GC'];

export const FALLBACK_FUTURES_ROOT = 'MES';

/** Derives a futures root from a chart or contract symbol; longest prefix wins. */
export function futuresRootFor(symbol: string): string | null {
  const uppercased = symbol.toUpperCase();
  return (
    [...KNOWN_FUTURES_ROOTS]
      .sort((a, b) => b.length - a.length)
      .find((root) => uppercased.startsWith(root)) ?? null
  );
}

File: apps/ios/0dteTrader/Features/Trade/FuturesRoots.swift
import Foundation

/// Known tradable futures roots.
enum FuturesRoots {
    static let known: [String] = ["MES", "ES", "MNQ", "NQ", "CL", "GC"]

    static let fallback = "MES"

    /// Derives a futures root from a chart symbol or contract symbol
    /// (e.g. "MES" → "MES", "MESU26" → "MES"). Longest prefixes first.
    static func root(for symbol: String) -> String? {
        let uppercased = symbol.uppercased()
        return known
            .sorted { $0.count > $1.count }
            .first { uppercased.hasPrefix($0) }
    }
}
